devito/ir/equations/algorithms.py

```diff
@@ -160,6 +157,30 @@ def _lower_exprs(expressions, subs):
         return processed.pop()
 
 
+def make_dimension_map(expr):
+    """
+    Make the dimension_map for an expression. In the basic case, this is extracted
+    directly from the SubDomain attached to the expression.
+
+    The indices of a Function defined on a SubDomain will all be the SubDimensions of
+    that SubDomain. In this case, the dimension_map should be extended with
+    `{ix_f: ix_i, iy_f: iy_i}` where `ix_f` is the SubDimension on which the Function is
+    defined, and `ix_i` is the SubDimension to be iterated over.
+    """
+    try:
+        dimension_map = {**expr.subdomain.dimension_map}
+
+        functions = [f for f in retrieve_functions(expr) if f._is_on_subdomain]
+        for f in functions:
+            dimension_map.update({d: expr.subdomain.dimension_map[d.root]
+                                  for d in f.space_dimensions if d.is_Sub})
+    except AttributeError:
+        # Some Relationals may be pure SymPy objects, thus lacking the SubDomain
+        dimension_map = {}
```

Comment:
Is this actually doing the right thing?
Looking at the previous code the `AttributeError` can only be raised in 1 place, here it can be raised in many. I'm not familiar enough with the code to know if this is safe.

I would have thought this should be something like:
```python
try:
    dimension_map = {**expr.subdomain.dimension_map}
except AttributeError:
    # Some Relationals may be pure SymPy objects, thus lacking the SubDomain
    dimension_map = {}
else:
    functions = [f for f in retrieve_functions(expr) if f._is_on_subdomain]
    for f in functions:
        dimension_map.update({d: expr.subdomain.dimension_map[d.root]
                              for d in f.space_dimensions if d.is_Sub})
```

Reply:
That would be safer. Changed.